devices/Atgm336h/Atgm336h.cs

```diff
@@ -0,0 +1,233 @@
+        }
+
+        private static Mode ConvertToMode(string data)
+        {
+            switch (data)
+            {
+                case "M":
+                    return Mode.Manual;
+                case "A":
+                    return Mode.Auto;
+            }
+
+            throw new Exception();
+        }
+
+        private static Fix ConvertToFix(string data)
+        {
+            switch (data)
+            {
+                case "1":
+                    return Fix.NoFix;
+                case "2":
+                    return Fix.Fix2D;
+                case "3":
+                    return Fix.Fix3D;
+            }
+
+            throw new Exception();
+        }
+        
+        private void SerialDevice_DataReceived(object sender, SerialDataReceivedEventArgs e)
+        {
+            var serialDevice = (SerialPort)sender;
+            if (serialDevice.BytesToRead == 0)
+            {
+                return;
+            }
+
+            var buffer = new byte[serialDevice.BytesToRead];
+            var bytesRead = serialDevice.Read(buffer, 0, buffer.Length);
+            var stringBuffer = Encoding.UTF8.GetString(buffer, 0, bytesRead);
+            var commands = stringBuffer.Split(serialDevice.WatchChar);
+            foreach (var command in commands)
+            {
+                var commandTrimmed = command.StartsWith("\n") ? command.Substring(1) : command;
+                if (commandTrimmed.StartsWith("$GNGSA"))
+                {
+                    var data = commandTrimmed.Split(',');
+                    var mode = ConvertToMode(data[1]);
+                    var fix = ConvertToFix(data[2]);
+
+                    if (fix != Fix)
+                    {
+                        Fix = fix;
+                        FixChanged?.Invoke(fix);
+                    }
+
+                    if (mode != Mode)
+                    {
+                        Mode = mode;
+                        ModeChanged?.Invoke(mode);
+                    }
+                }
+
+                if (Fix != Fix.NoFix && commandTrimmed.StartsWith("$GNGLL"))
+                {
+                    var data = commandTrimmed.Split(',');
+                    var lat = data[1];
+                    var latDir = data[2];
+                    var lon = data[3];
+                    var lonDir = data[4];
+                    Latitude = ConvertToGeoLocation(lat, latDir);
+                    Longitude = ConvertToGeoLocation(lon, lonDir);
+                    LocationChanged?.Invoke(Latitude, Longitude);
+                }
+            }
+        }
+        
+        /// <summary>
+        /// <inheritdoc/>
+        /// </summary>
+        public void Dispose()
+        {
+            if (_serialDevice is null)
+            {
+                return;
+            }
+            
+            if (_shouldDispose)
+            {
+                _serialDevice.Dispose();
+                _serialDevice.DataReceived -= SerialDevice_DataReceived;
```

Comment:
I would remove the event before disposing

Reply:
🤦‍♂️ 
Yep, you totally right.

Reply:
done